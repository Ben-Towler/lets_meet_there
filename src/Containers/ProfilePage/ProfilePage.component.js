import React from 'react';
import './ProfilePage.styles.scss';
import { firestore } from '../../Services/firebase.utils';






export default function ProfilePage({user}) {
  console.log(user);
  
  return (
    <div>
      {user && user.displayName}
    </div>
  )
}
