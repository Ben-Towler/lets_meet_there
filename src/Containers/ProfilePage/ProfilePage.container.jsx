/* eslint-disable react-hooks/exhaustive-deps */
import React, {useState, useEffect} from 'react';
import ApiClient from 'Services/ApiClient';
import { Trip, Spinner } from 'Components';
import { firestore } from 'Services/firebase.utils';
import helpers from '../../helpers';

export default function ProfilePage({user}) {
  const [isLoading, setIsLoading] = useState(false)
  const [favData, setFavData] = useState();

  const removeFromFavouritesHandler = async (origin, destination, outboundDate, inboundDate) => {
    const filteredData = user.favourites.filter(request => {
      const {userRequest} = request
      if (userRequest.origin === origin && userRequest.destination === destination && userRequest.outboundDate === outboundDate && userRequest.inboundDate === inboundDate) return false
      return true;
    });
    const userRef = await firestore.doc(`users/${user.id}`);
    userRef.update({ favourites: JSON.stringify([...filteredData]) });
  }
  

  function favListData() {
      return Promise.all(user.favourites.map(async requestData => {
        const {userRequest, friendRequest} = requestData;
        setIsLoading(true)
        const userFlightData = await ApiClient.getFavFlights(userRequest.origin, userRequest.destination, userRequest.outboundDate, userRequest.inboundDate )
        const friendFlightData = await ApiClient.getFavFlights(friendRequest.origin, friendRequest.destination, friendRequest.outboundDate, friendRequest.inboundDate)
        const userFlightList = userFlightData.quotes.sort((a,b) => a.MinPrice - b.MinPrice);
        const friendFlightList = friendFlightData.quotes.sort((a,b) => a.MinPrice - b.MinPrice);
        const userCityName = helpers.getCityName(userFlightData, userFlightList[0].OutboundLeg.OriginId)[0].CityName
        const friendCityName = helpers.getCityName(friendFlightData, friendFlightList[0].OutboundLeg.OriginId)[0].CityName
        const locationDetails = helpers.getLocation(userFlightData, userFlightList[0].OutboundLeg.DestinationId);

        return <Trip 
            key={Math.random() * 1000}
            favourites={true} 
            yourFlight={userFlightList[0]} 
            friendsFlight={friendFlightList[0]} 
            favLocation={{city: locationDetails[0].CityName, country: locationDetails[0].CountryName}} 
            userCity={userCityName} 
            friendCity={friendCityName} 
            removeFromFavouritesHandler={removeFromFavouritesHandler}
            searchDetailsForRemoveHandler= {userRequest}
            user={user}
          />
      }))
  }

  useEffect(() => {
    if (user.favourites && user.favourites.length ){ 
      favListData().then(data => {
        setIsLoading(false)
        setFavData(data);
      })
    }
  },[]);

  return (
    <React.Fragment>
      {
        isLoading 
          ? <Spinner /> 
          : <React.Fragment>
              {user && user.displayName}
              {user && user.favourites.length ? favData : <p id="noQuotes">You have no quotes favourited</p>}
            </React.Fragment>
      }
    </React.Fragment>
  )
}
