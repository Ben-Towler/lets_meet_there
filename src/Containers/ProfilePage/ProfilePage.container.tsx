import React, {useState, useEffect} from 'react';
import { useHistory } from 'react-router-dom'
import { getFavFlights } from 'Services/ApiClient';
import { Trip, Spinner } from 'Components';
import { firestore } from 'Services/firebase.utils';
import {User} from 'Interfaces/User'
import {FavFlight} from 'Interfaces/FavFlight';
import helpers from '../../helpers';

interface Props {
  user?: User;
}


export default function ProfilePage({user}: Props): JSX.Element {
  const [isLoading, setIsLoading] = useState(false)
  const [favData, setFavData] = useState<any>();
  const history = useHistory();

  const removeFromFavouritesHandler = async (origin: string, destination: string, outboundDate: string, inboundDate: string) => {
    
    const filteredData = user.favourites.filter(request => {
      const {userRequest} = request
      if (userRequest.origin === origin && userRequest.destination === destination && userRequest.outboundDate === outboundDate && userRequest.inboundDate === inboundDate) return false
      return true;
    });
    const userRef = await firestore.doc(`users/${user.id}`);
    userRef.update({ favourites: JSON.stringify([...filteredData]) });
  }
  

  function favListData() {
      return Promise.all(user.favourites.map(async requestData => {
        const {userRequest, friendRequest} = requestData;
        setIsLoading(true)
        const userFlightData = await getFavFlights(userRequest.origin, userRequest.destination, userRequest.outboundDate, userRequest.inboundDate )
        const friendFlightData = await getFavFlights(friendRequest.origin, friendRequest.destination, friendRequest.outboundDate, friendRequest.inboundDate)
        const userFlightList = userFlightData.quotes.sort((a: FavFlight,b: FavFlight) => a.MinPrice - b.MinPrice);
        const friendFlightList = friendFlightData.quotes.sort((a: FavFlight,b: FavFlight) => a.MinPrice - b.MinPrice);
        const userCityName = helpers.getCityName(userFlightData, userFlightList[0].OutboundLeg.OriginId)[0].CityName
        const friendCityName = helpers.getCityName(friendFlightData, friendFlightList[0].OutboundLeg.OriginId)[0].CityName
        const locationDetails = helpers.getLocation(userFlightData, userFlightList[0].OutboundLeg.DestinationId);

        return <Trip 
            key={Math.random() * 1000}
            favourites={true} 
            yourFlight={userFlightList[0]} 
            friendsFlight={friendFlightList[0]} 
            favLocation={{city: locationDetails[0].CityName, country: locationDetails[0].CountryName}} 
            userCity={userCityName} 
            friendCity={friendCityName} 
            removeFromFavouritesHandler={removeFromFavouritesHandler}
            searchDetailsForRemoveHandler= {userRequest}
            user={user}
          />
      }))
  }

  useEffect(() => {
    user || history.push('/')
    if (user && user.favourites && user.favourites.length ){ 
      favListData().then(data => {
        setIsLoading(false)
        setFavData(data);
      })
    }
  },[user]);

  return (
    <React.Fragment>
      {
        isLoading 
          ? <Spinner /> 
          : <React.Fragment>
              {user && user.displayName}
              {user && user.favourites && user.favourites.length ? favData : <p id="noQuotes">You have no quotes favourited</p>}
            </React.Fragment>
      }
    </React.Fragment>
  )
}

