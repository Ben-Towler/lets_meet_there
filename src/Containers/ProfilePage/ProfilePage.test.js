import React from 'react'
import {render} from '@testing-library/react'
import ProfilePage from '../ProfilePage/ProfilePage.container';
 
const mocks = {
  user: {
    displayName: 'Andrew',
    favourites: [{
      userRequest: {
        origin: 'LOND',
        destination: 'DUSS',
        outboundDate: '2020-06-23',
        inboundDate: '2020-06-25'
      },
      friendRequest: {
        origin: 'BERL',
        destination: 'DUSS',
        outboundDate: '2020-06-23',
        inboundDate: '2020-06-25'
      }
    }]
  },
  noDataUser: {
    displayName: 'Andrew',
    favourites: []
  }
};

describe('test the profile page load', () => {
  it('should match snapshot and render correctly', () => {
    const { asFragment } = render(<ProfilePage user={mocks.noDataUser} />);
    
    expect(asFragment(<ProfilePage />)).toMatchSnapshot();
  })
  
  it('should display user display name', () => {
    const { queryByText } = render(<ProfilePage user={mocks.noDataUser} />);
    const display = queryByText('Andrew');
    expect(display).toBeInTheDocument();
  })
  
  it('should tell the user when there are no favourite quotes', () => {
    const { queryByText } = render(<ProfilePage user={mocks.noDataUser} />);
    const noQuotes = queryByText('You have no quotes favourited');
    expect(noQuotes).toBeInTheDocument();
  })

  it('should render differently when favourite data exists', () => {
    const { queryByText } = render(<ProfilePage user={mocks.user} />);
    const noQuotes = queryByText('You have no quotes favourited');
    expect(noQuotes).not.toBeInTheDocument();
  })

})

