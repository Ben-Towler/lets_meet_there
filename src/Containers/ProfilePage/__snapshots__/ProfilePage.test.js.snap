// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the profile page load should match snapshot and render correctly 1`] = `
<DocumentFragment>
  Andrew
  <p
    id="noQuotes"
  >
    You have no quotes favourited
  </p>
</DocumentFragment>
`;
